platform/fabric/core/generic/msp/x509/loader.go

```diff
@@ -51,33 +55,99 @@ func GetSigningIdentity(mspConfigPath, mspID string) (SigningIdentity, error) {
 
 // LoadLocalMSPAt loads an MSP whose configuration is stored at 'dir', and whose
 // id and type are the passed as arguments.
-func LoadLocalMSPAt(dir, id, mspType string) (msp.MSP, error) {
+func LoadLocalMSPAt(dir, id, mspType string, bccspConfig *config.BCCSP) (msp.MSP, error) {
 	if mspType != "bccsp" {
 		return nil, errors.Errorf("invalid msp type, expected 'bccsp', got %s", mspType)
 	}
 	conf, err := msp.GetLocalMspConfig(dir, nil, id)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "could not get msp config from dir [%s]", dir)
 	}
-	ks, err := sw.NewFileBasedKeyStore(nil, filepath.Join(dir, "keystore"), true)
+
+	cp, _, err := GetBCCSPFromConf(dir, bccspConfig)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "failed to get bccsp from config [%v]", bccspConfig)
 	}
-	cryptoProvider, err := sw.NewDefaultSecurityLevelWithKeystore(sw.NewDummyKeyStore())
-	if err != nil {
-		return nil, err
+
+	mspOpts := &msp.BCCSPNewOpts{
+		NewBaseOpts: msp.NewBaseOpts{
+			Version: msp.MSPv1_0,
+		},
 	}
-	thisMSP, err := msp.NewBccspMspWithKeyStore(msp.MSPv1_0, ks, cryptoProvider)
+	thisMSP, err := msp.New(mspOpts, cp)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "failed to create new BCCSPMSP instance at [%s]", dir)
 	}
 	err = thisMSP.Setup(conf)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "failed to setup the new BCCSPMSP instance at [%s]", dir)
 	}
 	return thisMSP, nil
 }
 
+// GetBCCSPFromConf returns a BCCSP instance and its relative key store from the passed configuration.
+func GetBCCSPFromConf(dir string, conf *config.BCCSP) (bccsp.BCCSP, bccsp.KeyStore, error) {
+	if conf == nil {
+		return GetSWBCCSP(dir)
+	}
```

Comment:
oh this should return an error, no?

Reply:
it does.

Reply:
no, I was asking if this function should return an error if no config is provided (`conf == nil`)?!?!!?! In that situation, the current impl returns `GetSWBCCSP`. If this is the preferred default behaviour, please add it to the function go doc.

Reply:
okay